tiledb/sm/rtree/rtree.cc

```diff
@@ -95,62 +89,85 @@ RTree& RTree::operator=(RTree&& rtree) noexcept {
 /*               API              */
 /* ****************************** */
 
+Status RTree::build_tree() {
+  if (levels_.empty())
+    return Status::Ok();
+
+  assert(levels_.size() == 1);
+
+  auto leaf_num = levels_[0].size();
+  if (leaf_num == 1)
+    return Status::Ok();
+
+  // Build the tree bottom up
+  auto height = (size_t)ceil(utils::math::log(fanout_, leaf_num)) + 1;
+  for (size_t i = 0; i < height - 1; ++i) {
+    auto new_level = build_level(levels_.back());
+    levels_.emplace_back(new_level);
+  }
+
+  // Make the root as the first level
+  std::reverse(std::begin(levels_), std::end(levels_));
+
+  return Status::Ok();
+}
+
 unsigned RTree::dim_num() const {
-  return dim_num_;
+  return (domain_ == nullptr) ? 0 : domain_->dim_num();
+}
+
+const Domain* RTree::domain() const {
+  return domain_;
 }
 
 unsigned RTree::fanout() const {
   return fanout_;
 }
 
-template <class T>
-TileOverlap RTree::get_tile_overlap(const std::vector<const T*>& range) const {
+TileOverlap RTree::get_tile_overlap(const NDRange& range) const {
   TileOverlap overlap;
 
   // Empty tree
-  if (dim_num_ == 0 || levels_.empty())
+  if (domain_ == nullptr || levels_.empty())
     return overlap;
 
   // This will keep track of the traversal
   std::list<Entry> traversal;
   traversal.push_front({0, 0});
-  auto leaf_num = levels_.back().mbr_num_;
+  auto leaf_num = levels_.back().size();
   auto height = this->height();
-  uint64_t mbr_size = 2 * dim_num_ * datatype_size(type_);
 
   while (!traversal.empty()) {
     // Get next entry
     auto entry = traversal.front();
     traversal.pop_front();
-    auto level = entry.level_;
-    auto mbr_idx = entry.mbr_idx_;
-    auto offset = entry.mbr_idx_ * mbr_size;
-    auto mbr = levels_[level].mbrs_.data() + offset;
+    const auto& mbr = levels_[entry.level_][entry.mbr_idx_];
 
-    // Get
-    auto ratio = this->range_overlap<T>(range, (T*)mbr);
+    // Get overlap ratio
+    auto ratio = domain_->overlap_ratio(range, mbr);
 
     // If there is overlap
     if (ratio != 0.0) {
```

Comment:
As I alluded to in an above comment, I don't think 'ratio' will ever be equal to 0.

Reply:
See above.